File: app/src/main/java/com/micewine/emu/utils/FilePathResolver.java
package com.micewine.emu.utils;

import static com.micewine.emu.activities.MainActivity.wineDisksFolder;

import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.util.Log;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class FilePathResolver {
    public static String resolvePath(Context context, Uri uri) {
        String realPath = resolveRealPathFromUri(context, uri);

        if (realPath != null) {
            File[] wineDisks = wineDisksFolder.listFiles();
            if (wineDisks == null) return realPath;

            try {
                for (File disk : wineDisks) {
                    String symlinkTo = disk.getCanonicalPath();
                    if (realPath.startsWith(symlinkTo)) {
                        return realPath.replaceFirst(symlinkTo, disk.getPath());
                    }
                }
            } catch (Exception ignored) {
            }
        }

        return realPath;
    }

    /*
    private fun getDosHomeDirs(): List<String> {
        val dosHomeDirs = mutableListOf<String>()
        val dosHomeDir = File(MainActivity.fileManagerDefaultDir)

        if (dosHomeDir.exists() && dosHomeDir.isDirectory) {
            dosHomeDirs.addAll(dosHomeDir.listFiles()?.map { it.absolutePath } ?: emptyList())
        }

        return dosHomeDirs
    }
     */

    private static String resolveRealPathFromUri(Context context, Uri uri) {
        if (DocumentsContract.isDocumentUri(context, uri)) {
            if (isExternalStorageDocument(uri)) {
                String docId = DocumentsContract.getDocumentId(uri);
                String[] docSplit = docId.split(":");
                String type = docSplit[0];

                if ("primary".equalsIgnoreCase(type)) {
                    if (docSplit.length > 1) {
                        return Environment.getExternalStorageDirectory().toString() + "/" + docSplit[1];
                    } else {
                        return Environment.getExternalStorageDirectory().toString() + "/";
                    }
                } else {
                    File file = new File("storage" + "/" + docId.replace(":", "/"));
                    if (file.exists()) {
                        return "/storage/" + docId.replace(":", "/");
                    }
                    List<String> availableExternalStorages = getStorageDirectories();
                    String root = "";

                    for (String storagePath : availableExternalStorages) {
                        if (docSplit[1].startsWith("/")) {
                            root = storagePath + docSplit[1];
                        } else {
                            root = storagePath + "/" + docSplit[1];
                        }
                    }
                    if (root.contains(type)) {
                         return "storage" + "/" + docId.replace(":", "/");
                    } else {
                        if (root.startsWith("/storage/") || root.startsWith("storage/")) {
                            return root;
                        } else if (root.startsWith("/")) {
                            return "/storage" + root;
                        } else {
                            return "/storage/" + root;
                        }
                    }
                }
            } else if (isRawDownloadsDocument(uri)) {
                String fileName = getFilePath(context, uri);
                String subFolderName = getSubFolders(uri);
                if (fileName != null) {
                    return Environment.getExternalStorageDirectory().toString() + "/Download/" + subFolderName + fileName;
                }
                String docId = DocumentsContract.getDocumentId(uri);
                Uri contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), Long.parseLong(docId));

                return getDataColumn(context, contentUri, null, null);
            } else if (isDownloadsDocument(uri)) {
                String fileName = getFilePath(context, uri);

                if (fileName != null) {
                    return Environment.getExternalStorageDirectory().toString() + "/Download/" + fileName;
                }

                String docId = DocumentsContract.getDocumentId(uri);

                if (docId.startsWith("raw:")) {
                    docId = docId.replaceFirst("raw:", "");
                    if (new File(docId).exists()) return docId;
                }
                if (docId.startsWith("raw%3A%2F")) {
                    docId = docId.replaceFirst("raw%3A%2F", "");
                    if (new File(docId).exists()) return docId;
                }
                Uri contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), Long.parseLong(docId));

                return getDataColumn(context, contentUri, null, null);
            } else if (isMediaDocument(uri)) {
                String docId = DocumentsContract.getDocumentId(uri);
                String[] docSplit = docId.split(":");
                String selection = "_id=?";
                String[] selectionArgs = new String[] { docSplit[1] };

                return getDataColumn(context, null, selection, selectionArgs);
            }
        } else if ("content".equalsIgnoreCase(uri.getScheme())) {
            String dataColumn = getDataColumn(context, uri, null, null);
            if (dataColumn == null) {
                Log.e("FileResolver", "Failed to resolve path from uri.");
            }
            return dataColumn;
        } else if ("file".equalsIgnoreCase(uri.getScheme())) {
            return uri.getPath();
        }

        return null;
    }

    private static String getSubFolders(Uri uri) {
        String replaceChars = uri.toString()
                .replace("%2F", "/")
                .replace("%20", " ")
                .replace("%3A", ":");

        String[] bits = replaceChars.split("/");

        int size = bits.length;
        if (size < 6) {
            return "";
        }

        String sub5 = bits[size - 2];
        String sub4 = bits[size - 3];
        String sub3 = bits[size - 4];
        String sub2 = bits[size - 5];
        String sub1 = bits[size - 6];

        if ("Download".equals(sub1)) {
            return sub2 + "/" + sub3 + "/" + sub4 + "/" + sub5 + "/";
        } else if ("Download".equals(sub2)) {
            return sub3 + "/" + sub4 + "/" + sub5 + "/";
        } else if ("Download".equals(sub3)) {
            return sub4 + "/" + sub5 + "/";
        } else if ("Download".equals(sub4)) {
            return sub5 + "/";
        } else {
            return "";
        }
    }

    private static String getDataColumn(Context context, Uri uri, String selection, String[] selectionArgs) {
        Cursor cursor = null;
        String column = "_data";
        String[] projection = new String[] { column };

        try {
            cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs, null);

            if (cursor != null && cursor.moveToFirst()) {
                int index = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(index);
            }
        } catch (Exception ignored) {
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        return null;
    }

    private static String getFilePath(Context context, Uri uri) {

        String[] projection = new String[] { MediaStore.Files.FileColumns.DISPLAY_NAME };
        try (Cursor cursor = context.getContentResolver().query(uri, projection, null, null, null)) {
            if (cursor != null && cursor.moveToFirst()) {
                int index = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DISPLAY_NAME);
                return cursor.getString(index);
            }
        } catch (Exception ignored) {
        }

        return null;
    }

    private static boolean isExternalStorageDocument(Uri uri) {
        return "com.android.externalstorage.documents".equals(uri.getAuthority());
    }

    private static boolean isDownloadsDocument(Uri uri) {
        return "com.android.providers.downloads.documents".equals(uri.getAuthority());
    }

    private static boolean isRawDownloadsDocument(Uri uri) {
        return uri.toString().contains("com.android.providers.downloads.documents/document/raw");
    }

    private static boolean isMediaDocument(Uri uri) {
        return "com.android.providers.media.documents".equals(uri.getAuthority());
    }

    private static List<String> getStorageDirectories() {
        List<String> paths = new ArrayList<>();
        paths.add(Environment.getExternalStorageDirectory().toString()); // Adiciona o caminho padrão do armazenamento externo
        paths.addAll(getAdditionalStoragePaths());
        return paths;
    }

    private static List<String> getAdditionalStoragePaths() {
        List<String> storagePaths = new ArrayList<>();
        File file = new File("/storage");

        if (file.exists()) {
            File[] files = file.listFiles();
            if (files != null) {
                for (File f : files) {
                    storagePaths.add(f.getAbsolutePath());
                }
            }
        }

        return storagePaths;
    }

}

File: app/src/main/java/com/micewine/emu/utils/FilePathResolver.kt
package com.micewine.emu.utils

import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore
import android.util.Log
import com.micewine.emu.activities.MainActivity
import java.io.File

object FilePathResolver {
    fun resolvePath(context: Context, uri: Uri): String? {
        val realPath = resolveRealPathFromUri(context, uri) // Get the resolved real path

        if (realPath != null) {
            getDosHomeDirs().forEach { dosHome ->
                val symlinkTo = File(dosHome).canonicalPath
                if (realPath.startsWith(symlinkTo)) {
                    val resolvedPath = realPath.replaceFirst(symlinkTo, dosHome)
                    return resolvedPath
                }
            }
        }

        return realPath
    }

    private fun getDosHomeDirs(): List<String> {
        val dosHomeDirs = mutableListOf<String>()
        val dosHomeDir = File(MainActivity.fileManagerDefaultDir)

        if (dosHomeDir.exists() && dosHomeDir.isDirectory) {
            dosHomeDirs.addAll(dosHomeDir.listFiles()?.map { it.absolutePath } ?: emptyList())
        }

        return dosHomeDirs
    }

    private fun resolveRealPathFromUri(context: Context, uri: Uri): String? {
        if (DocumentsContract.isDocumentUri(context, uri)) {
            if (isExternalStorageDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":").toTypedArray()
                val type = split[0]
                return if ("primary".equals(type, ignoreCase = true)) {
                    if (split.size > 1) {
                        Environment.getExternalStorageDirectory().toString() + "/" + split[1]
                    } else {
                        Environment.getExternalStorageDirectory().toString() + "/"
                    }
                } else {
                    if (File("storage" + "/" + docId.replace(":", "/")).exists()) {
                        return "/storage/" + docId.replace(":", "/")
                    }
                    val availableExternalStorages = getStorageDirectories()
                    var root = ""
                    for (s in availableExternalStorages) {
                        root = if (split[1].startsWith("/")) {
                            s + split[1]
                        } else {
                            s + "/" + split[1]
                        }
                    }
                    if (root.contains(type)) {
                        "storage" + "/" + docId.replace(":", "/")
                    } else {
                        if (root.startsWith("/storage/") || root.startsWith("storage/")) {
                            root
                        } else if (root.startsWith("/")) {
                            "/storage$root"
                        } else {
                            "/storage/$root"
                        }
                    }
                }
            } else if (isRawDownloadsDocument(uri)) {
                val fileName = getFilePath(context, uri)
                val subFolderName = getSubFolders(uri)
                if (fileName != null) {
                    return Environment.getExternalStorageDirectory().toString() + "/Download/" + subFolderName + fileName
                }
                val id = DocumentsContract.getDocumentId(uri)
                val contentUri = ContentUris.withAppendedId(
                    Uri.parse("content://downloads/public_downloads"),
                    java.lang.Long.valueOf(id)
                )

                return getDataColumn(context, contentUri, null, null)
            } else if (isDownloadsDocument(uri)) {
                val fileName = getFilePath(context, uri)

                if (fileName != null) {
                    return Environment.getExternalStorageDirectory().toString() + "/Download/" + fileName
                }
                var id = DocumentsContract.getDocumentId(uri)

                if (id.startsWith("raw:")) {
                    id = id.replaceFirst("raw:".toRegex(), "")
                    val file = File(id)
                    if (file.exists()) return id
                }
                if (id.startsWith("raw%3A%2F")) {
                    id = id.replaceFirst("raw%3A%2F".toRegex(), "")
                    val file = File(id)
                    if (file.exists()) return id
                }
                val contentUri = ContentUris.withAppendedId(
                    Uri.parse("content://downloads/public_downloads"),
                    java.lang.Long.valueOf(id)
                )

                return getDataColumn(context, contentUri, null, null)
            } else if (isMediaDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":").toTypedArray()
                val selection = "_id=?"
                val selectionArgs = arrayOf(split[1])

                return getDataColumn(context, null, selection, selectionArgs)
            }
        } else if ("content".equals(uri.scheme, ignoreCase = true)) {
            if (getDataColumn(context, uri, null, null) == null) {
                Log.d("FileResolver", "Failed to resolve path from uri: dataReturnedNull")
            }
            return getDataColumn(context, uri, null, null)
        } else if ("file".equals(uri.scheme, ignoreCase = true)) {
            return uri.path
        }
        return null
    }

    private fun getSubFolders(uri: Uri): String {
        val replaceChars = uri.toString().replace("%2F", "/").replace("%20", " ").replace("%3A", ":")
        val bits = replaceChars.split("/").toTypedArray()
        val sub5 = bits[bits.size - 2]
        val sub4 = bits[bits.size - 3]
        val sub3 = bits[bits.size - 4]
        val sub2 = bits[bits.size - 5]
        val sub1 = bits[bits.size - 6]

        return when {
            sub1 == "Download" -> "$sub2/$sub3/$sub4/$sub5/"
            sub2 == "Download" -> "$sub3/$sub4/$sub5/"
            sub3 == "Download" -> "$sub4/$sub5/"
            sub4 == "Download" -> "$sub5/"
            else -> ""
        }
    }

    private fun getDataColumn(
        context: Context,
        uri: Uri?,
        selection: String?,
        selectionArgs: Array<String>?
    ): String? {
        var cursor: Cursor? = null
        val column = "_data"
        val projection = arrayOf(column)
        try {
            cursor =
                context.contentResolver.query(uri!!, projection, selection, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val index = cursor.getColumnIndexOrThrow(column)
                return cursor.getString(index)
            }
        } catch (e: Exception) {
            Log.d("FileResolver", "Failed to resolve path from uri: ${e.message}")
        } finally {
            cursor?.close()
        }
        return null
    }

    private fun getFilePath(context: Context, uri: Uri?): String? {
        var cursor: Cursor? = null
        val projection = arrayOf(MediaStore.Files.FileColumns.DISPLAY_NAME)
        try {
            cursor = context.contentResolver.query(
                uri!!, projection, null, null,
                null
            )
            if (cursor != null && cursor.moveToFirst()) {
                val index = cursor.getColumnIndexOrThrow(MediaStore.Files.FileColumns.DISPLAY_NAME)
                return cursor.getString(index)
            }
        } catch (e: Exception) {
            Log.d("FileResolver", "Failed to resolve path from uri: ${e.message}")
        } finally {
            cursor?.close()
        }
        return null
    }

    private fun isExternalStorageDocument(uri: Uri): Boolean {
        return "com.android.externalstorage.documents" == uri.authority
    }

    private fun isDownloadsDocument(uri: Uri): Boolean {
        return "com.android.providers.downloads.documents" == uri.authority
    }

    private fun isRawDownloadsDocument(uri: Uri): Boolean {
        val uriToString = uri.toString()
        return uriToString.contains("com.android.providers.downloads.documents/document/raw")
    }

    private fun isMediaDocument(uri: Uri): Boolean {
        return "com.android.providers.media.documents" == uri.authority
    }

    private fun getStorageDirectories(): List<String> {
        val paths = mutableListOf<String>()
        paths.add(Environment.getExternalStorageDirectory().toString()) // Add the default external storage path
        paths.addAll(getAdditionalStoragePaths())

        return paths
    }

    private fun getAdditionalStoragePaths(): List<String> {
        val storagePaths = mutableListOf<String>()
        val file = File("/storage")

        if (file.exists()) {
            file.listFiles()?.forEach { storagePaths.add(it.absolutePath) }
        }

        return storagePaths
    }
}
